[SR-14761] Improve diagnostic for implicitly overridden init()

HannaYakusevych opened the pull request:
Provide more detailed diagnostic message when implicitly synthesized @objc initializer in subclass was overridden inside the subclass' extension

Resolves SR-14761

xedin reviewed (approved):
Makes sense to me. @CodaFi Could you please take a look?

xedin commented on lib/Sema/TypeCheckDeclPrimary.cpp:
```diff
@@ -627,7 +627,8 @@ CheckRedeclarationRequest::evaluate(Evaluator &eval, ValueDecl *current) const {
     const auto *otherOverride = other->getOverriddenDecl();
-    if (currentOverride && currentOverride == otherOverride) {
+    const auto *otherInit = dyn_cast<ConstructorDecl>(other);
+    if (currentOverride && currentOverride == otherOverride && !(otherInit && otherInit->isImplicit())) {
```
One nit: This doesn't look like it has been formatted by clang-format. I'd suggest you use utility from llvm-project repo - `llvm-project/clang/tools/clang-format`, you just need to add that to your PATH and run it after `git add`. I think some IDEs also support that as a formatter...

HannaYakusevych replied:
Done

HannaYakusevych committed:
[SE-14761] Changed diagnostic for implicitly overridden init()

Provide more detailed diagnostic message when implicitly synthesized @objc initializer in subclass was overridden inside the subclass' extension

Resolves: rdar://problem/79179597

CodaFi commented:
Looks good

@swift-ci smoke test

xedin commented:
@swift-ci please smoke test Linux platform

xedin commented:
@swift-ci please clean smoke test Linux platform

xedin commented:
@swift-ci please clean smoke test Linux platform

xedin commented:
@swift-ci please clean smoke test Linux platform

xedin commented:
@swift-ci please clean smoke test Linux platform

xedin commented:
@swift-ci please smoke test Linux platform
